sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/QueryUtils.java

```diff
@@ -0,0 +1,311 @@
+    }
+
+    return additionalOrders;
+  }
+
+  private static void fillInequalityFields(Filter filter, List<String> result) {
+    switch (filter.getFilterTypeCase()) {
+      case FIELD_FILTER:
+        if (INEQUALITY_FIELD_FILTER_OPS.contains(filter.getFieldFilter().getOp())) {
+          String fieldPath = filter.getFieldFilter().getField().getFieldPath();
+          if (!result.contains(fieldPath)) {
+            result.add(fieldPath);
+          }
+        }
+        break;
+      case COMPOSITE_FILTER:
+        filter.getCompositeFilter().getFiltersList().forEach(f -> fillInequalityFields(f, result));
+        break;
+      case UNARY_FILTER:
+        if (INEQUALITY_UNARY_FILTER_OPS.contains(filter.getUnaryFilter().getOp())) {
+          String fieldPath = filter.getUnaryFilter().getField().getFieldPath();
+          if (!result.contains(fieldPath)) {
+            result.add(fieldPath);
+          }
+        }
+        break;
+      default:
+        break;
+    }
+  }
+
+  @Nullable
+  static Value lookupDocumentValue(List<String> segments, Map<String, Value> valueMap) {
+    String field = segments.remove(0);
+    Value value = valueMap.get(field);
+    if (segments.isEmpty()) {
+      return value;
+    }
+    // Field path traversal is not done, recurse into map values.
+    if (value == null || !value.getValueTypeCase().equals(ValueTypeCase.MAP_VALUE)) {
+      return null;
+    }
+    return lookupDocumentValue(segments, value.getMapValue().getFieldsMap());
+  }
+
+  private static final String UNQUOTED_NAME_REGEX_STRING = "([a-zA-Z_][a-zA-Z_0-9]*)";
+  private static final String QUOTED_NAME_REGEX_STRING = "(`(?:[^`\\\\]|(?:\\\\.))+`)";
+  // After each segment follows a dot and more characters, or the end of the string.
+  private static final Pattern FIELD_PATH_SEGMENT_REGEX =
+      Pattern.compile(
+          String.format("(?:%s|%s)(\\..+|$)", UNQUOTED_NAME_REGEX_STRING, QUOTED_NAME_REGEX_STRING),
+          Pattern.DOTALL);
+
+  static List<String> resolveOrderByFieldPath(String fieldPath) throws IllegalArgumentException {
+    if (fieldPath.isEmpty()) {
+      throw new IllegalArgumentException("Could not resolve empty field path");
+    }
+    List<String> segments = new ArrayList<>();
+    while (!fieldPath.isEmpty()) {
+      Matcher segmentMatcher = FIELD_PATH_SEGMENT_REGEX.matcher(fieldPath);
+      boolean foundMatch = segmentMatcher.lookingAt();
+      if (!foundMatch) {
+        throw new IllegalArgumentException("OrderBy field path was malformed");
+      }
+      String fieldName = segmentMatcher.group(1);
+      // Unquoted group is null, use quoted group.
+      if (fieldName == null) {
+        fieldName = segmentMatcher.group(2);
+        String escaped = escapeFieldName(fieldName.substring(1, fieldName.length() - 1));
+        segments.add(escaped);
+      } else {
+        segments.add(fieldName);
+      }
+      fieldPath = fieldPath.substring(fieldName.length());
+      if (fieldPath.startsWith(".")) {
```

Comment:
Because the regex matched, fieldPath is either empty here, or it is `.nestedfieldname`, so I'm removing the dot. `if(!fieldPath.isEmpty())` would have the same effect, but I think it's probably less clear?

Reply:
SG, maybe add a inline comment that when fieldPath is not empty it must start with "." here.

Reply:
Done.